content/docs/reference/self-hosted/authentication-provider/bring-your-own.md

```diff
@@ -0,0 +1,124 @@
+---
+title: Bring your own Authentication Provider
+id: '/docs/reference/self-hosted/authentication-provider/bring-your-own'
+prev: '/docs/reference/self-hosted/authentication-provider/tina-cloud'
+next: null
+---
+
+To self host your own authentication, you will need to do do two things.
+
+## 1. Create a custom Authentication Provider Class
+
+You can do so by extending the `AbstractAuthProvider` class and implementing the following functions:
+
+### Required Functions
+
+`authenticate`: This function is called when the user goes into `/admin` and they are not logged in (determined by `getUser`). This function should redirect the user to the login page or do whatever is necessary to authenticate the user.
+
+`getUser`: This function is called when the user goes into `/admin` and is used to determine if the user is logged in. If it returns a truthy value, the user is logged in and if it returns a falsy value the user is not logged in.
+
+`getToken`: This function is called when a request is made to the GraphQL endpoint. It should return an object with an `id_token` property. This will be passed as an `Authorization` header in the format `Bearer <id_token>`
+
+`logOut`: This function is called when the user clicks the logout button in the admin.
+
+### Optional Functions
+
+`getSessionProvider`: Return a React context provider to that will wrap the admin
+
+```ts
+import { AbstractAuthProvider } from 'tinacms'
+
+export class CustomAuthProvider extends AbstractAuthProvider {
+  constructor() {
+    super()
+    // Do any setup here
+  }
+  async authenticate(props?: {}): Promise<any> {
+    // Do any authentication here
+  }
+  getToken() {
+    // Return the token here. The token will be passed as an Authorization header in the format `Bearer <token>`
+  }
+  async getUser() {
+    // Returns a truthy value, the user is logged in and if it returns a falsy value the user is not logged in.
+  }
+  logout() {
+    // Do any logout logic here
+  }
+  async authorize(context?: any): Promise<any> {
+    // Do any authorization logic here
+  }
+  getSessionProvider() {
```

Comment:
I think we can leave `getSessionProvider` in here I will just make the comment more explicit that it can be deleted if its not needed. 

There may be other auth providers that need it.